server/mqtt.go

```diff
@@ -1338,30 +1373,48 @@ func (s *Server) mqttCreateAccountSessionManager(acc *Account, quitCh chan struc
 			return nil, fmt.Errorf("failed to update stream config: %w", err)
 		}
 	}
-	// Try to transfer regardless if we have already updated the stream or not
-	// in case not all messages were transferred and the server was restarted.
-	if as.transferRetainedToPerKeySubjectStream(s) {
+
+	transferRMS := func() error {
+		if !needToTransfer {
+			return nil
+		}
+
+		as.transferRetainedToPerKeySubjectStream(s)
+
 		// We need another lookup to have up-to-date si.State values in order
 		// to load all retained messages.
 		si, err = lookupStream(mqttRetainedMsgsStreamName, "retained messages")
 		if err != nil {
-			return nil, err
+			return err
 		}
+		needToTransfer = false
+		return nil
 	}
+
+	// Attempt to transfer all "single subject" retained messages to new
+	// subjects. It may fail, will log its own error; ignore it the first time
+	// and proceed to updating MaxMsgsPer. Then we invoke transferRMS() again,
+	// which will get another chance to resolve the error; if not we bail there.
+	transferRMS()
```

Comment:
We must check the error here, and return if not nil. The reason is that if the transfer is ok, but the lookup failed, the `si` state would be wrong. I guess that since original si.Config.MaxMsgsPer would likely be != 1, we would update the stream (which updates `si`) and we do check for error at the last call to transferRMS(). Still, I think it would be correct to return error if lookup failed here.

Reply:
But if we return the error here, we would then get into the risky situation you were describing earlier, no? I thought we were trying to apply the MaxPerSubject update even on transfer error (we'd _always_ do it if we are transferring); the update would then reduce the "old" messages to 1, and then the second attempt will either clear it successfully (and update si), or fail altogether.

Reply:
Also, maybe there is a way to delete that last remaining message without loading it first, I just don't know how; just deleting it would probably be cleaner (2nd pass).

Reply:
Different that what we discussed. In this case, it is the lookupStream only that can return an error. That I would be fine with the idea that if we can't lookup the stream, we ought to fail. More over, if we have a failure looking up the stream:
```
		si, err = lookupStream(mqttRetainedMsgsStreamName, "retained messages")
		if err != nil {
			return err
		}
```
then `si` will be set to `nil`. So in that case we would get a panic at the next call:
```
	if si.Config.MaxMsgsPer != 1 {
		si.Config.MaxMsgsPer = 1
...
```
since `si` would have been set to nil due to failed `lookupStream()` call.

Reply:
You're right. Updated both to return on error.